<!doctype html>
<html>
	<head>
		<meta charset="utf-8">
		<meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no">

		<title>scRNA-Seq</title>

		<link rel="stylesheet" href="reveal.js/dist/reset.css">
		<link rel="stylesheet" href="reveal.js/dist/reveal.css">
		<link rel="stylesheet" href="reveal.js/dist/theme/white.css" id="theme">
		<link rel="stylesheet" href="reveal.js/dist/user.css">

		<!-- Theme used for syntax highlighted code -->
		<link rel="stylesheet" href="reveal.js/plugin/highlight/monokai.css" id="highlight-theme">
		<style type="text/css">
			.inline {
				background-color: #272822;
				color: white;
			}
		</style>
	</head>
	<body>
		<div class="reveal">
			<div class="slides">
				<section>
					<h1>Single cell RNA-Seq analysis</h1>
					<p>Bastian Schiffthaler</p>
				</section>
				<section>
					<section data-background="white">
						<h3>Workflow</h3>
						<img class="stretch" data-src="https://raw.githubusercontent.com/Bioconductor/OSCABase/images/images/Workflow.png" ></img>
						<p class="small"><a href="https://bioconductor.org/books/release/OSCA/overview.html">https://bioconductor.org/books/release/OSCA/overview.html</a></p>
					</section>
					<section>
						<h3>Technologies</h3>
						<ul>
							<li class="fragment">Droplet-based: 10X Genomics, inDrop, Drop-seq</li>
							<li class="fragment">Plate-based with unique molecular identifiers (UMIs): CEL-seq, MARS-seq</li>
							<li class="fragment">Plate-based with reads: Smart-seq2</li>
							<li class="fragment">Other: sci-RNA-seq, Seq-Well</li>
						</ul>
					</section>
					<section>
						<h3>Evolution of scRNA Seq</h3>
						
						<img class="stretch" data-src="https://raw.githubusercontent.com/bschiffthaler/scRNA-Seq/main/img/41596_2018_Article_BFnprot2017149_Fig1_HTML.webp"></img>
					</section>
					<section>
						<h3>Technologies: comparison</h3>
						<img class="stretch" data-src="https://els-jbs-prod-cdn.jbs.elsevierhealth.com/cms/attachment/303d109c-368c-42ef-aaec-2a66c7908df6/gr2_lrg.jpg"></img>
					</section>
					<section>
						<h3>Smart-Seq</h3>
						<div style="display: flex;">
						<ul>
							<li class="fragment">RNA-Seq + cell sorting</li>
							<li class="fragment">Essentially developed to use for tiny RNA amounts</li>
							<li class="fragment">Cannot use UMI</li>
						</ul>
						<img class="fragment" height="50%" data-src="img/smart.webp"></img>
						</div>
					</section>
					<section>
						<h3>Drop-Seq</h3>
						<img class="stretch" data-src="https://els-jbs-prod-cdn.jbs.elsevierhealth.com/cms/attachment/b00311ef-6cbb-48eb-b042-646efb9c64bb/gr2_lrg.jpg"></img>
						<p class="small">Macosko, Evan Z., et al. "Highly parallel genome-wide expression profiling of individual cells using nanoliter droplets." Cell 161.5 (2015): 1202-1214.</p>
					</section>
					<section>
						<h3>Demultiplexing doublets using genotype</h3>
						<img class="stretch" data-src="https://raw.githubusercontent.com/bschiffthaler/scRNA-Seq/main/img/demug_gt.webp">
						<p class="small">Kang, H., Subramaniam, M., Targ, S. et al. Multiplexed droplet single-cell RNA-sequencing using natural genetic variation. Nat Biotechnol 36, 89â€“94 (2018). </p>
					</section>
					<section>
						<h3>Demultiplexing doublets using antibodies</h3>
						<img class="stretch" data-src="https://raw.githubusercontent.com/bschiffthaler/scRNA-Seq/main/img/hashdemux.webp">
						<p class="small">Stoeckius, M., Zheng, S., Houck-Loomis, B. et al. Cell Hashing with barcoded antibodies enables multiplexing and doublet detection for single cell genomics. Genome Biol 19, 224 (2018) </p>
					</section>
					<section>
						<h3>10X</h3>
						<img class="stretch" data-src="https://raw.githubusercontent.com/bschiffthaler/scRNA-Seq/main/img/10x.webp">
						<p class="small">Zheng, G., Terry, J., Belgrader, P. et al. Massively parallel digital transcriptional profiling of single cells. Nat Commun 8, 14049 (2017).</p>
					</section>
					<section>
						<h3>Combinatorial Indexing</h3>
						<img class="stretch" data-src="https://raw.githubusercontent.com/bschiffthaler/scRNA-Seq/main/img/ci.jpeg">
						<p class="small">Cao, Junyue, et al. "Comprehensive single-cell transcriptional profiling of a multicellular organism." Science 357.6352 (2017)</p>
					</section>
				</section>
				<section>
					<section>
						<h3>Preprocessing</h3>
						<ul>
							<li class="fragment">Read based (Smart-Seq2)?: Same as bulk RNA-Seq</li>
							<li class="fragment">10X?: Cellranger suite or...</li>
							<li class="fragment"><code class="inline">salmon alevin</code></li>
						</ul>
					</section>
					<section>
						<h3><code class="inline bash">salmon alevin</code></h3>
						<ul><li class="fragment">Alevin has many relevant options for various technologies</li></ul>
						<pre class="fragment"><code style="max-height: 800px !important;" data-trim data-noescape class="text">
						  --dropseq                             Use DropSeq Single Cell protocol for
						                                        the library
						  --chromiumV3                          Use 10x chromium v3 Single Cell
						                                        protocol for the library.
						  --chromium                            Use 10x chromium v2 Single Cell
						                                        protocol for the library.
						  --gemcode                             Use 10x gemcode v1 Single Cell protocol
						                                        for the library.
						  --citeseq                             Use CITESeq Single Cell protocol for
						                                        the library, 16 CB, 12 UMI and
						                                        features.
						  --celseq                              Use CEL-Seq Single Cell protocol for
						                                        the library.
						  --celseq2                             Use CEL-Seq2 Single Cell protocol for
						                                        the library.
						  --quartzseq2                          Use Quartz-Seq2 v3.2 Single Cell
						                                        protocol for the library assumes 15
						                                        length barcode and 8 length UMI.
						</code></pre>
					</section>
					<section>
						<h3>The R1 read contains cell barcode and UMI</h3>
						<pre><code data-trim data-noescape class="text">
							 @D000684:779:H53GNBCXY:1:1101:1198:2139 1:N:0:AAATGTGC
							 TNTCCTCGTTACCGAT..TGCATGCCCG

							[----------------][__________]
							 CCCCCCCCCCCCCCCC  UUUUUUUUUU
						</code></pre>
					</section>
				</section>
				<section>
					<section>
						<h3>Data import and QC</h3>
						<ul>
							<li class="fragment" data-fragment-index=1><code class="inline">DropletUtils</code> - Cellranger</li>
							<li class="fragment" data-fragment-index=2><code class="inline">tximport</code> - Salmon alevin</li>
							<li class="fragment" data-fragment-index=3><code class="inline">tximeta</code> - Salmon alevin</li>
						</ul>
						<br><br><hr class="fragment" data-fragment-index=4>
						<p class="fragment" data-fragment-index=4>Tximeta/Tximport:</p>
						<pre class="fragment"><code class="r" data-fragment-index=4 data-trim data-noescape>
							library(tximeta)
							library(SingleCellExperiment)

							sc &lt;- tximeta(in_dir, "alevin", alevinArgs=list(filterBarcodes=TRUE))
							sce &lt;- as(sc, "SingleCellExperiment")
						</code></pre>
						<p class="fragment" data-fragment-index=4>DropletUtils:</p>
						<pre class="fragment"><code class="r" data-fragment-index=4 data-trim data-noescape>
							library(DropletUtils)

							sce &lt;- read10xCounts(in_dir)
						</code></pre>
					</section>
					<section>
						<h3>SingleCellExperiments</h3>
						<img class="stretch" data-src="https://raw.githubusercontent.com/Bioconductor/OSCABase/images/images/SingleCellExperiment.png"></img>
					</section>
					<section>
						<h3>QC metrics</h3>
						<ul>
							<li class="fragment">Cells with small library sizes</li>
							<li class="fragment">Cells with few expressed genes (diversity)</li>
							<li class="fragment">High proportion of reads mapped to spike-ins</li>
							<li class="fragment">High proportion of reads mapped to mitochondrial genes**</li>
							<li class="fragment">Empty cells (droplets)</li>
							<li class="fragment">Doublets (droplets)</li>
						</ul>

						<p class="small fragment">** High proportions are indicative of poor-quality cells (Islam et al. 2014; Ilicic et al. 2016), presumably because of loss of cytoplasmic RNA from perforated cells. The reasoning is that, in the presence of modest damage, the holes in the cell membrane permit efflux of individual transcript molecules but are too small to allow mitochondria to escape, leading to a relative enrichment of mitochondrial transcripts. For single-nuclei RNA-seq experiments, high proportions are also useful as they can mark cells where the cytoplasm has not been successfully stripped</p>
					</section>
					<section>
						<h3>Droplet specific - empty cells</h3>
						<ul>
							<li>Not all empty cells are free of RNA</li>
							<li><code class="inline">DropletUtils</code> can use simulations to detect</li>
						</ul>
						<pre><code class="r" data-trim>
							empt <- emptyDrops(counts(sce))
						</code></pre>
						<img class="stretch" data-src="https://bioconductor.org/books/release/OSCA.advanced/droplet-processing_files/figure-html/rankplot-1.png"></img>
					</section>
					<section>
						<h3>Droplet specific - doublets</h3>
						<ul>
							<li> Cluster based: check for clusters whose expression lies between two other clusters</li>
							<li> Simulation based</li>
						</ul>
						<pre><code class="r" data-trim>
							library(scDblFinder)
							dbl.out <- findDoubletClusters(sce)
							dbl.dens <- computeDoubletDensity(sce, d=ncol(reducedDim(sce)))
							plotTSNE(sce, colour_by="DoubletScore")
						</code></pre>
						<img class="stretch" data-src="https://bioconductor.org/books/release/OSCA.advanced/doublet-detection_files/figure-html/denstsne-1.png"></img>
					</section>
					<section>
						<h3>Getting QC metrics</h3>
						<pre><code class="r" data-trim>
							# Extract mito sequences
							mito <- seqnames(rowRanges(sc)) == "chrM"
							sc_m <- sc[mito, ]
							# Create SCE dataset with mitos as subset
							sce <- as(sc, "SingleCellExperiment")
							altExp(sce, "mito") <- sc_m
							# Add QC metrics
							sce <- addPerCellQC(sce)
						</code></pre>
					</section>
					<section>
						<h3>Check if discarded outliers are confounded with biology</h3>
						<img class="stretch" data-src="https://bioconductor.org/books/release/OSCA.basic/quality-control_files/figure-html/qc-dist-416b-1.png"></img>
					</section>
					<section>
						<h3>Check if discarded outliers are confounded with biology (2)</h3>
						<img class="stretch" data-src="https://bioconductor.org/books/release/OSCA.basic/quality-control_files/figure-html/qc-mito-zeisel-1.png"></img>
					</section>
					<section>
						<h3>Check if discarded outliers are confounded with biology (3)</h3>
						<img class="stretch" data-src="https://bioconductor.org/books/release/OSCA.basic/quality-control_files/figure-html/qc-mito-spike-zeisel-1.png"></img>
					</section>
					<!-- <section>
						<h3>Check if discarded outliers are confounded with biology (4)</h3>
						<div style="display: flex;">
						<img width="50%" flex="1" data-src="https://bioconductor.org/books/release/OSCA.basic/quality-control_files/figure-html/discardplot416b-1.png"></img>
						<img width="50%" flex="1" data-src="https://bioconductor.org/books/release/OSCA.basic/quality-control_files/figure-html/discardplotpbmc-1.png"></img>
						</div>
					</section> -->
				</section>
				<section>
					<section>
						<h3>Normalization by size factors</h3>
						<ul>
							<li class="fragment">Size factor (library size) - not actually balanced hence inaccurate for scRNA Seq</li>
						</ul>
						<pre class="fragment"><code class="r" data-trim>
							library_sf <- librarySizeFactors(sce)
						</code></pre>
						<blockquote class="fragment">
							In practice, normalization accuracy is not a major consideration for exploratory scRNA-seq data analyses. Composition biases do not usually affect the separation of clusters, only the magnitude - and to a lesser extent, direction - of the log-fold changes between clusters or cell types. As such, library size normalization is usually sufficient in many applications where the aim is to identify clusters and the top markers that define each cluster.
						</blockquote>
					</section>
					<section>
						<h3>Normalization by Deconvolution</h3>
						<ul>
						<li class="fragment">Normalization by deconvolution<ol>
								<li class="fragment">Cluster similar cells</li>
								<li class="fragment">Pool cells to reduce dropouts</li>
								<li class="fragment">Deconvolve size factors by cluster (cell type)</li>
							</ol></li>
						</ul>
						<pre class="fragment"><code class="r" data-trim>
							clust <- quickCluster(sce)
							deconv_sf <- calculateSumFactors(sce, cluster=clust)
						</code></pre>
					</section>
					<section>
						<img class="stretch" data-src="https://bioconductor.org/books/release/OSCA.basic/normalization_files/figure-html/deconv-zeisel-1.png"></img>
					</section>
					<section>
						<h3>Spike ins</h3>
						<pre class="fragment"><code class="r" data-trim>
							sce.richard <- computeSpikeFactors(sce.richard, "ERCC")
						</code></pre>
					</section>
				</section>
				<section>
					<section>
						<h3>Feature selection</h3>
						<ul>
							<li class="fragment">Not all variation is interesting (biological)</li>
							<li class="fragment">Extract genes with most promising variability</li>
							<li class="fragment">Model technical variation, then extract most relevant genes</li>
							<li class="fragment">Use either mean-variance model, or coefficient of variation</li>
						</ul>
					</section>
					<section>
						<pre><code class="r" data-trim>
							dec <- modelGeneVar(sce)
							fit <- metadata(dec)
							plot(fit$mean, fit$var, xlab="Mean of log-expression",
							    ylab="Variance of log-expression")
							curve(fit$trend(x), col="dodgerblue", add=TRUE, lwd=2)
						</code></pre>
						<img class="stretch" data-src="https://bioconductor.org/books/release/OSCA.basic/feature-selection_files/figure-html/trend-plot-pbmc-1.png"></img>
					</section>
					<section>
						<pre><code class="r" data-trim>
							cv2 <- modelGeneCV2(sce)
							fit2 <- metadata(cv2)
							plot(fit2$mean, fit2$cv2, log="xy", xlab="Mean of log expression",
							     ylab="log coefficient of variation ^ 2",
							     xlim = c(1e-05, 1e2))
							curve(fit2$trend(x), col="dodgerblue", add=TRUE, lwd=2)
						</code></pre>
						<img class="stretch" data-src="https://bioconductor.org/books/release/OSCA.advanced/feature-selection_files/figure-html/cv2-pbmc-1.png"></img>
					</section>
				</section>
				<section>
					<section>
						<h3>Dimensionality reduction techniques</h3>
						<ul>
							<li class="fragment">Aim: Visualize and analyze high dimensional data in lower dimensions</li>
							<li class="fragment">PCA</li>
							<li class="fragment">t-SNE</li>
							<li class="fragment">UMAP</li>
						</ul>
					</section>
					<section>
						<h3>Principal Component Analysis</h3>
						<img class="stretch" data-src="https://bioconductor.org/books/release/OSCA.basic/reduced-dimensions_files/figure-html/zeisel-pca-multi-1.png"></img>
					</section>
					<section>
						<h3>t-SNE</h3>
						<img class="stretch" data-src="https://bioconductor.org/books/release/OSCA.basic/reduced-dimensions_files/figure-html/tsne-brain-1.png"></img>
					</section>
					<section>
						<h3>UMAP</h3>
						<img class="stretch" data-src="https://bioconductor.org/books/release/OSCA.basic/reduced-dimensions_files/figure-html/umap-brain-1.png"></img>
					</section>
					<section>
						<h3>Criticisms</h3>
						<img class="stretch" data-src="https://raw.githubusercontent.com/bschiffthaler/scRNA-Seq/main/img/crit.png"></img>
						<p class="small">https://www.biorxiv.org/content/10.1101/2021.08.25.457696v2</p>
					</section>
				</section>
				<section>
					<section>
						<h3>Clustering</h3>
						<ul>
							<li class="fragment">Aim: Assign each cell to a group</li>
							<li class="fragment">Graph-based: walktrap, infomap</li>
							<li class="fragment">Known number: k-means</li>
						</ul>
					</section>
					<section>
						<h3>Graph based methods</h3>
						<ol>
							<li class="fragment">Convert relationship into graph by nearest neighbours</li>
							<li class="fragment">Apply clustering algorithm</li>
						</ol>
					</section>
					<section>
						<pre><code class="r" data-trim>
							library(scran)
							library(igraph)
							g <- buildSNNGraph(sce, k=10, use.dimred = 'PCA')
							clust <- cluster_walktrap(g)$membership
							colLabels(sce) <- factor(clust)
							plotReducedDim(sce, "TSNE", colour_by="label")
						</code></pre>
						<img class="stretch" data-src="https://bioconductor.org/books/release/OSCA.basic/clustering_files/figure-html/tsne-clust-graph-1.png"></img>
					</section>
				</section>
				<section>
					<section>
						<h3>Marker gene detection</h3>
						<ul>
							<li class="fragment">Detect genes active (DE-up) in clusters</li>
							<li class="fragment">Pairwise comparison of all clusters</li>
							<li class="fragment">Options: <ul>
								<li class="fragment">t-test (p): difference of means</li>
								<li class="fragment">wilcoxon test (AUC): optimal separation</li>
								<li class="fragment">binomial test: proportion of cells expressing</li>
								<li class="fragment">DESeq2, edgeR, voom-limma, ...</li>
							</ul></li>
						</ul>
					</section>
					<section>
						<pre><code class="r" data-trim>
							markers <- findMarkers(sce)
							interesting <- rownames(markers[[7]][1:5, ])
							logFCs <- getMarkerEffects(interesting)
							pheatmap(logFCs, breaks=seq(-5, 5, length.out=101))
						</code></pre>
						<img class="stretch" data-src="https://bioconductor.org/books/release/OSCA.advanced/marker-detection_files/figure-html/heat-focused-pbmc-1.png"></img>
					</section>
					<section>
						<pre><code class="r" data-trim>
							plotExpression(sce, x="label", features=interesting)
						</code></pre>
						<img class="stretch" data-src="https://bioconductor.org/books/release/OSCA.advanced/marker-detection_files/figure-html/viol-de-binom-1.png"></img>
					</section>
				</section>
				<section>
					<section>
						<h3>Cell type annotation</h3>
						<ul>
							<li class="fragment">Often bottleneck, depends on a priori knowledge</li>
							<li class="fragment">Not perfectly well defined</li>
							<li class="fragment">Options:<ul>
								<li class="fragment">GO cluster enrichment (and interpretation)</li>
								<li class="fragment">KEGG cluster enrichment (and interpretation)</li>
								<li class="fragment">Reference data set</li>
							</ul></li>
						</ul>
					</section>
					<section>
						<pre><code class="r" data-trim>
							ref <- BlueprintEncodeData()
							pred <- SingleR(test=sce, ref=ref, labels=ref$label.main)
							table(pred$labels)
							plotScoreHeatmap(pred)
						</code></pre>
						<img class="stretch" data-src="https://bioconductor.org/books/release/OSCA.basic/cell-annotation_files/figure-html/singler-heat-pbmc-1.png"></img>
					</section>
					<section>
						<pre><code class="r" data-trim>
							sce$label <- pred$labels
 							plotReducedDim(sce, "UMAP", colour_by = "label")
						</code></pre>
						<img class="stretch" data-src="https://raw.githubusercontent.com/bschiffthaler/scRNA-Seq/main/img/umap_ctype.png"></img>
					</section>
				</section>
			</div>
		</div>

		<script src="reveal.js/dist/reveal.js"></script>
		<script src="reveal.js/plugin/notes/notes.js"></script>
		<script src="reveal.js/plugin/markdown/markdown.js"></script>
		<script src="reveal.js/plugin/highlight/highlight.js"></script>
		<script src="reveal.js/plugin/math/math.js"></script>
		<script>
			// More info about initialization & config:
			// - https://revealjs.com/initialization/
			// - https://revealjs.com/config/
			Reveal.initialize({
				hash: true,
				width: 1600,
				height: 1280,

				// Learn about plugins: https://revealjs.com/plugins/
				plugins: [ RevealMarkdown, RevealHighlight, RevealNotes, RevealMath ]
			});
		</script>
	</body>
</html>
